dataplane/kernel-forwarder/pkg/kernelforwarder/kernelforwarder.go

```diff
@@ -0,0 +1,140 @@
+
+package kernelforwarder
+
+import (
+	"context"
+
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/opentracing/opentracing-go"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/grpc/status"
+
+	"github.com/networkservicemesh/networkservicemesh/dataplane/pkg/common"
+
+	"github.com/networkservicemesh/networkservicemesh/controlplane/pkg/apis/crossconnect"
+	local "github.com/networkservicemesh/networkservicemesh/controlplane/pkg/apis/local/connection"
+	remote "github.com/networkservicemesh/networkservicemesh/controlplane/pkg/apis/remote/connection"
+	"github.com/networkservicemesh/networkservicemesh/dataplane/pkg/apis/dataplane"
+	"github.com/networkservicemesh/networkservicemesh/pkg/tools"
+)
+
+// KernelForwarder instance
+type KernelForwarder struct {
+	common *common.DataplaneConfig
+}
+
+// CreateKernelForwarder creates an instance of the KernelForwarder
+func CreateKernelForwarder() *KernelForwarder {
+	return &KernelForwarder{}
+}
+
+// MonitorMechanisms handler
+func (v *KernelForwarder) MonitorMechanisms(empty *empty.Empty, updateSrv dataplane.Dataplane_MonitorMechanismsServer) error {
+	logrus.Infof("MonitorMechanisms was called")
+	initialUpdate := &dataplane.MechanismUpdate{
+		RemoteMechanisms: v.common.Mechanisms.RemoteMechanisms,
+		LocalMechanisms:  v.common.Mechanisms.LocalMechanisms,
+	}
+	logrus.Infof("Sending MonitorMechanisms update: %v", initialUpdate)
+	if err := updateSrv.Send(initialUpdate); err != nil {
+		logrus.Errorf("Kernel forwarding plane server: Detected error %s, grpc code: %+v on grpc channel", err.Error(), status.Convert(err).Code())
+		return nil
+	}
+	// Waiting for any updates which might occur during a life of dataplane module and communicating
+	// them back to NSM.
+	for update := range v.common.MechanismsUpdateChannel {
+		v.common.Mechanisms = update
+		logrus.Infof("Sending MonitorMechanisms update: %v", update)
+		if err := updateSrv.Send(&dataplane.MechanismUpdate{
+			RemoteMechanisms: update.RemoteMechanisms,
+			LocalMechanisms:  update.LocalMechanisms,
+		}); err != nil {
+			logrus.Errorf("Kernel forwarding plane server: Detected error %s, grpc code: %+v on grpc channel", err.Error(), status.Convert(err).Code())
+			return nil
+		}
+	}
+	return nil
+}
+
+// Request handler for connections
+func (v *KernelForwarder) Request(ctx context.Context, crossConnect *crossconnect.CrossConnect) (*crossconnect.CrossConnect, error) {
+	logrus.Infof("Request() called with %v", crossConnect)
+	xcon, err := v.connectOrDisconnect(crossConnect, cCONNECT)
+	if err != nil {
+		return nil, err
+	}
+	v.common.Monitor.Update(xcon)
+	logrus.Infof("Request() called with %v returning: %v", crossConnect, xcon)
+	return xcon, err
+}
+
+func (v *KernelForwarder) connectOrDisconnect(crossConnect *crossconnect.CrossConnect, connect bool) (*crossconnect.CrossConnect, error) {
+	/* 1. Handle local connection */
+	if crossConnect.GetLocalSource().GetMechanism().GetType() == local.MechanismType_KERNEL_INTERFACE &&
```

Comment:
If memif will go here, we will see message about invalid remote mechanism, I suppose we need to move check from handleRemoteConnection and put it on last.

Reply:
Well, the Kernel Forwarder does not announce itself as capable of handling MemIf interfaces. It should not be called with MemIf.

Reply:
Indeed, we don't advertise local connection support for Memif 👍

Reply:
Although this should be handled by the NSM manager, I'll put an additional sanity check to see whether the request type is indeed one of the supported ones we advertise when the forwarding plane is registered 👍